### PR TITLE
[ISSUE-2742] A lighter InsertPlan for single data point insertion

### PR DESCRIPTION
**Pull request by WilliamSong112**
## Description
Hi all :
I have submitted a pr for [ISSUE-2742](https://github.com/apache/iotdb/issues/2742), please take a look.
This issue about add the number of  insert data point by point into a single time series.  He was able to avoid some of the shortcomings of Insertplan (a large number of array objects need to be constructed)

The modules involved are as follows:
   1、add a new physical plan <InsertSinglePointPlan>
   2、Added two new methods to session . insertSinglePoint  insertStringSinglePoint
Here are some of his integration tests :
image

<hr>

This PR has:
- [x] been self-reviewed.
    - [x] concurrent read
    - [x] concurrent write
    - [x] concurrent read and write 
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. 
- [x] added or updated version, __license__, or notice information
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [x] added integration tests.
- [x] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: docs/zh/UserGuide/API/Programming-Java-Native-API.md [hunk 1]
```diff
@@ -158,6 +158,18 @@ void insertRecord(String deviceId, long time, List<String> measurements,
    List<TSDataType> types, List<Object> values)
 ```
 
+* 插入单个点，一个 Record 是一个设备一个时间戳下单个测点的数据。提供数据类型后，服务器不需要做类型推断，可以提高性能
+
+```java
+void insertSinglePoint(String deviceId, long time, String measurement, TSDataType type, String value)
+```
+
+* 插入单个点，一个 Record 是一个设备一个时间戳下单个测点的数据。提供数据类型后，服务器不需要做类型推断，可以提高性能
```
Comment (HTHou) [inline docs/zh/UserGuide/API/Programming-Java-Native-API.md:167]:
```suggestion
* 插入单个点，一个 Record 是一个设备一个时间戳下单个测点的数据。服务器需要做类型推断，可能会有额外耗时
```

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java [hunk 2]
```diff
@@ -366,6 +415,87 @@ private void checkMemCostAndAddToTspInfo(InsertTabletPlan insertTabletPlan, int
     updateMemoryInfo(memTableIncrement, chunkMetadataIncrement, textDataIncrement);
   }
 
+  private void checkMemCostAndAddToTspInfo(InsertSinglePointPlan insertSinglePointPlan)
+      throws WriteProcessException {
```
Comment (HTHou) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:419]: Remove this?

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java [hunk 3]
```diff
@@ -366,6 +415,87 @@ private void checkMemCostAndAddToTspInfo(InsertTabletPlan insertTabletPlan, int
...
+    }
+    // TEXT data mem size
+    if (insertSinglePointPlan.getDataType() == TSDataType.TEXT) {
+      textDataIncrement += MemUtils.getBinarySize((Binary) insertSinglePointPlan.getValue());
+    }
+    updateMemoryInfo(memTableIncrement, chunkMetadataIncrement, textDataIncrement);
+  }
+
+  private void checkMemcheckMemCostAndAddToTspInfoCostAndAddToTspInfo(
+      InsertTabletPlan insertTabletPlan, int start, int end) throws WriteProcessException {
```
Comment (HTHou) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:460]: Remove this?

Comment (WilliamSong112) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java]: Thank you for your reply，This function is independent of insert single point。I've removed it

File: docs/zh/UserGuide/API/Programming-Java-Native-API.md (hunk 1 above)
Comment (WilliamSong112) [reply docs/zh/UserGuide/API/Programming-Java-Native-API.md]: Thank you for your reply，This documentation error has been fixed

**Comment by jixuan1989** [comment]
Hi, can you give some experimental results to show the performance enhancement?

File: server/src/main/java/org/apache/iotdb/db/metadata/MManager.java [hunk 4]
```diff
@@ -2013,6 +2014,19 @@ public void collectMeasurementSchema(
     }
   }
 
+  private TSDataType getType(InsertSinglePointPlan plan) throws MetadataException {
```
[the remark this answers is not in the document]
Comment (WilliamSong112) [reply server/src/main/java/org/apache/iotdb/db/metadata/MManager.java:2017]: > Hi, can you give some experimental results to show the performance enhancement?

Thank you for your reply，I know what you mean，I'll think about it，Refer to existing test experiments or existing 
 benchmark

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 5]
```diff
@@ -831,6 +832,43 @@ public void insert(InsertRowPlan insertRowPlan)
     }
   }
 
+  public void InsertSinglePoint(InsertSinglePointPlan insertSinglePointPlan)
```
Comment (jixuan1989) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:835]: 1. this function has too many duplicate codes with `insert()`;
2. Trigger related methods must also be maintained.

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 6]
```diff
@@ -1087,6 +1125,38 @@ private void insertToTsFileProcessor(
     }
   }
 
+  private void insertSinglePointToTsFileProcessor(
```
Comment (jixuan1989) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:1128]: the same issue, too many duplicated codes.

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java [hunk 7]
```diff
@@ -216,6 +218,53 @@ public void insert(InsertRowPlan insertRowPlan) throws WriteProcessException {
     tsFileResource.updatePlanIndexes(insertRowPlan.getIndex());
   }
 
+  /**
+   * insert data in an InsertRowPlan into the workingMemtable.
+   *
+   * @param insertSinglePointPlan physical plan of insertion
+   */
+  public void InsertSinglePoint(InsertSinglePointPlan insertSinglePointPlan)
```
Comment (jixuan1989) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:226]: the same issue, too many duplicated codes.

**Review by jixuan1989 — CHANGES_REQUESTED** [review]
Overall, there are too many duplicated codes.
You may have to refactor the InsertPlan, and use abstract class to reduce duplicated codes.

**Pull request closed without merging**